examples/tv-casting-app/linux/main.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/server/Mdns.h>
+#include <app/server/Server.h>
+#include <controller/CHIPCommissionableNodeController.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConfigurationManager.h>
+#include <system/SystemLayer.h>
+#include <transport/raw/PeerAddress.h>
+
+using namespace chip;
+using namespace chip::Controller;
+
+#define TV_DEVICE_TYPE 35
+
+CommissionableNodeController commissionableNodeController;
+chip::System::SocketWatchToken token;
+
+/**
+ * Enters commissioning mode, opens commissioning window, logs onboarding payload.
+ * If non-null selectedCommissioner is provided, sends user directed commissioning
+ * request to the selectedCommissioner and advertises self commissionable node over DNS-SD
+ */
+void PrepareForCommissioning(const Mdns::DiscoveredNodeData * selectedCommissioner = nullptr)
+{
+    // Enter commissioning mode, open commissioning window
+    InitServer();
+    ReturnOnFailure(OpenBasicCommissioningWindow(ResetFabrics::kYes, 3 * 60));
+
+    // Display onboarding payload
+    chip::DeviceLayer::ConfigurationMgr().LogDeviceConfig();
+
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    if (selectedCommissioner != nullptr)
+    {
+        // Advertise self as Commissionable Node over mDNS
+        ReturnOnFailure(app::Mdns::AdvertiseCommissionableNode(app::Mdns::CommissioningMode::kEnabledBasic));
+
+        // Send User Directed commissioning request
+        ReturnOnFailure(SendUserDirectedCommissioningRequest(chip::Transport::PeerAddress::UDP(
+            selectedCommissioner->ipAddress[0], selectedCommissioner->port, selectedCommissioner->interfaceId[0])));
+    }
+#endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+}
+
+/**
+ * Accepts user input of selected commissioner and calls PrepareForCommissioning with
+ * the selected commissioner
+ */
+void RequestUserDirectedCommisisioning(System::SocketEvents events, intptr_t data)
+{
+    // Accept user selection for commissioner to request commissioning from
+    int selectedCommissionerNumber = CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES;
+    scanf("%d", &selectedCommissionerNumber);
```

Comment:
What happens here if data not available due to non-blocking? Does non-blocking stdin actually work against stdin? This will unblock on socket watching, but scanf() may block if there had been insufficient input. Is it `\n` that unblocks stdin in this case?

Reply:
I had the same question.  Yes, we could use raw mode and poll each character.  But I'm pretty sure that the kernel will have line buffering in most environments by default, so in a reasonable majority of cases stdin will only unblock when the user actually presses enter.

Reply:
OK, then I suggest making that assumption (obviously made here) clear with a comment.

Reply:
Unresolving this.. I really don't think we should land examples that block the main thread with interactive input.

Reply:
I see that has been addressed although it it not obvious and I think there are still issues with it

- direct scanf won't well with more than one command
- end of file results in an infinite loop except on a tty (e.g. suppose this example is run as a subprocess with input redirected. It will not function)

Overall I wouldn't recommend this approach. If we want line editing in examples we should do it via shell_common.